third-party/realdds/src/dds-adapter-watcher.cpp

```diff
@@ -0,0 +1,158 @@
+#include <rsutils/signal.h>
+#include <rsutils/easylogging/easyloggingpp.h>
+#include <rsutils/time/stopwatch.h>
+
+#include <fastrtps/utils/IPFinder.h>
+using eprosima::fastrtps::rtps::IPFinder;
+
+
+namespace realdds {
+
+
+using ip_set = dds_adapter_watcher::ip_set;
+
+
+namespace detail {
+
+
+class adapter_watcher_singleton
+{
+    std::shared_ptr< rsutils::os::adapter_watcher > _adapter_watcher;
+
+    ip_set _ips;
+
+    std::thread _th;
+    rsutils::time::stopwatch _time_since_update;
+
+    using public_signal
+        = rsutils::public_signal< adapter_watcher_singleton, ip_set const & /*new*/, ip_set const & /*old*/ >;
+
+public:
+    public_signal callbacks;
+
+    adapter_watcher_singleton()
+        : _adapter_watcher( std::make_shared< rsutils::os::adapter_watcher >(
+            [this]()
+            {
+                // Adapters have changed, but we won't see the effects for some time (several seconds)
+                // But we don't want to wait here, so we start another thread
+                _time_since_update.reset();
+                if( ! _th.joinable() )
+                {
+                    _th = std::thread(
+                        [this, weak = std::weak_ptr< rsutils::os::adapter_watcher >( _adapter_watcher )]
+                        {
+                            LOG_DEBUG( "waiting for IP changes" );
+                            ip_set new_ips, old_ips;
+                            while( auto strong = weak.lock() )
+                            {
+                                if( _time_since_update.get_elapsed() > std::chrono::seconds( 15 ) )
+                                    break;
+                                std::this_thread::sleep_for( std::chrono::seconds( 1 ) );
+                                new_ips.clear();
+                                old_ips.clear();
+                                update_ips( &new_ips, &old_ips );
+                                if( new_ips.size() || old_ips.size() )
+                                    callbacks.raise( new_ips, old_ips );
+                            }
+                            _th.detach();  // so it's not joinable
+                            LOG_DEBUG( "done waiting for IP changes" );
+                        } );
+                }
+            } ) )
+    {
+        LOG_DEBUG( "network adapter watcher singleton is up" );
+        update_ips();
+    }
+
+    ~adapter_watcher_singleton()
+    {
+        _adapter_watcher.reset();  // signal the thread to finish
+        if( _th.joinable() )
+            _th.join();
+    }
+
+    void update_ips( ip_set * p_new_ips = nullptr, ip_set * p_old_ips = nullptr )
+    {
+        auto ips = dds_adapter_watcher::current_ips();
+
+        auto first1 = ips.begin();
+        auto const last1 = ips.end();
+        auto first2 = _ips.begin();
+        auto const last2 = _ips.end();
+
+        while( first1 != last1 )
+        {
+            if( first2 == last2  || *first1 < *first2 )
```

Comment:
This is an implementation of a set-difference function.
We proceed alphabetically because that's how the items are kept, and because we need to (efficiently) know when an item from one set does not appear in the other.

Reply:
You need to see what first1 and first2 are assigned to to understand which is new/old

Reply:
Renamed `first1` -> `new_it`
Renamed `last1` -> `new_end`
Renamed `first2` -> `old_it`
Renamed `last2` -> `old_end`

Reply:
Now I see
I missed the _ difference so it looked the same.